### PR TITLE
feat: added CMMD metric

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Tests

### Description

Adds CMMD metric.
Updates `.coreset` `@property` to work with supervised coresets -> fixes the `compute_metric` method.
### How Has This Been Tested?

Added unit tests.

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: coreax/metrics.py [hunk 1]
```diff
@@ -129,3 +132,141 @@ def compute(
             self.precision_threshold,
         )
         return jnp.sqrt(squared_mmd_threshold_applied)
+
+
+class CMMD(Metric):
```
Comment (tc85324) [inline coreax/metrics.py:137]: Missing the specification of the generic. I think the `MMD` metric is wrong with its specification, all `_Data` should be replaced with `Data` in `MMD`. Similarly, `_SupervisedData` with `SupervisedData` in CMMD, unless there is a requirement to maintain a TypeVar?

File: coreax/metrics.py [hunk 2]
```diff
@@ -129,3 +132,141 @@ def compute(
...
+
+        inverse_feature_gramian_2 = inverse_approximator.approximate(
+            array=feature_gramian_2,
+            regularisation_parameter=self.regularisation_parameter,
+            identity=jnp.eye(feature_gramian_2.shape[0]),
+        )
+
+        # Compute each term in the CMMD
+        term_1 = (
+            inverse_feature_gramian_1
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/metrics.py:248]: Is there any way to leverage the blocked computation, via `compute_mean` or otherwise, like in `MMD`?

**Review by tc85324 — COMMENTED** [review]
Like the addition of this metric! My concerns are primarily with the additions to `coreset.py`.  As far as I can see, these could be removed entirely from this pull and/or left for future discussion (excluding the change to `Coreset.coreset` which is neccessary)?

File: coreax/metrics.py [hunk 3]
```diff
@@ -129,3 +132,141 @@ def compute(
...
+        )
+        term_3 = (
+            inverse_feature_gramian_1
+            @ self.response_kernel.compute(y1, y2)
+            @ inverse_feature_gramian_2
+            @ self.feature_kernel.compute(x2, x1)
+        )
+
+        # Compute CMMD
+        squared_cmmd = jnp.trace(term_1) + jnp.trace(term_2) - 2 * jnp.trace(term_3)
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:264]: Given symmetric matrices $A, B \in \mathbb{R}^{n\times n}$, we have $\text{Tr}(AB) := \sum_{i=1}^n(AB)_{ii} = \sum_{i,j=1}^n a_{ij}{bji} = \sum_{i,j=1}^n  (A \cdot B)_{ij}$ i.e. the sum of their Hadamard product. So we could compute the trace this way with no wasted work. But I remember testing this previously and not seeing any time improvement. Will double check and report back

File: coreax/metrics.py (hunk 1 above)
Comment (db091756) [reply coreax/metrics.py:137]: Fixed

File: coreax/metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:245]: Yep. 

I don't think we can use the blocked computation in any meaningful way as we are never computing a mean here as far as I can see. We could perhaps block this process by doing block matrix inversion but we would need to instantiate the entire feature gramian anyway at a memory cost that is equivalent to what we would be trying to save.

File: coreax/coreset.py [hunk 4]
```diff
@@ -122,12 +124,43 @@ class Coresubset(Coreset[_Data], Generic[_Data]):
...
+        """
+        Return the coresubset with the order of its coreset indices permuted.
+
+        Convenience method for changing the order of coreset index refinement.
+        """
+        return Coresubset(
+            Data(jr.permutation(random_key, self.unweighted_indices)),
+            self.pre_coreset_data,
+        )
+
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coreset.py:165]: These methods would replace the old functionality of `RefineReverse` and add additional functionality to randomise order of refinement.  Because the methods in  `RefinementSolver.refine(...)` work in the order that the coreset indices are inputted you can get stuck in local minima where additional recursive calls to refinement are stuck. By reversing or permuting the order of the coreset indices you can jump out of these local minima at no cost.

Thanks for the general spots!

File: coreax/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:264]: Top is current way, bottom is Hadamard way 
image
Does seem like the Hadamard way saves time. Not sure what happened last time!

Added.

File: coreax/coreset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coreset.py:165]: I'm getting pylint complaints with Self but not with "Coresubset"

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coreset.py]: image

File: coreax/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/metrics.py]: This is a nice win!

File: coreax/metrics.py (hunk 1 above)
Comment (tc85324) [reply coreax/metrics.py]: Sorry, I should have made it clearer. The  `class MMD(Metric[_Data])` should also be to `MMD(Metric[Data])`, similarly for the supervised data cases. This means you can remove the `_SupervisedData` type var completely.

Comment (db091756) [reply coreax/metrics.py]: Fixed properly now! See above for issues with the Hadamard change

File: coreax/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/metrics.py:264]: Interesting, can you post benchmark times for the complete CMMD metric computation (when JIT compiled) for both approaches, and with 32bit and 64bit default precision. I'll have a quick think why this might be.

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:264]: Elaborating on the above, when $A, B$ are rectangular matrices, we have
```math
\text{Tr}(A^TB) := \sum_{i=1}^{n} (A^TB)_{ii} = \sum_{i=1}^{n} \sum_{m=1}^{n}a_{ij}b_{ij} = \sum_{i,j=1}^{n}  (A^T \odot B)_{ij}
```

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:264]: We could just compute the third term as a trace but this doesn't remove the numerical instability issues that the tests are flagging i.e. computing the CMMD on a dataset with itself seems much more likely to be non-zero or negative with the Hadamard method based on some fiddling I've done.

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:264]: Will do! In the office for a team day today so not sure how much coding I will be able to get done 

> Interesting, can you post benchmark times for the complete CMMD metric computation (when JIT compiled) for both approaches, and with 32bit and 64bit default precision. I'll have a quick think why this might be.

Will do! In the office today for a team day so not sure how much coding I will be doing.

File: coreax/coreset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/coreset.py]: I think the type hinting issue can be fixed by using `eqx.tree_at`. This way the generic types will be preserved and the type hint complain should go away. In general, I would prefer the user to be responsible for performing these index re-orderings as It only requires making a copy of the coreset with updated nodes. For example 
```python
reversed_coreset = eqx.tree_at(lambda x: x.nodes, coreset, coreset.nodes[::-1])
permuted_indices = jr.choice(key, len(coreset))
permuted_coreset = eqx.tree_at(lambda x: x.nodes, coreset, coreset.nodes[permuted_indices])
```
This allows the user to perform arbitrary re-indexing of the `nodes`, which is more powerful than what is exposed by the convenience interfaces, and avoids any ambiguity about what exactly is being permuted or reversed.

I'll add that from a design standpoint, the responsibility of handling the problem of local-minima should be that of the refinement solver itself, or better yet, some "controller" that wraps/interfaces with the solver. Unless there is an obvious demand from users to have these convenience interfaces, I would argue that the more appropriate solution here would be to provide a means to properly control the refinement solvers to avoid local-minima (or at least to enumerate a set of local-minima). This would certainly require a separate ticket though.

File: coreax/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/metrics.py:263]: If you group the matrix multiplies as so: `jnp.sum((A @ B).T * (C @ D))`, using the brackets to ensure the matmul is before the element wise multiply, do we see any appreciable changes in speed? Regarding the shapes, aren't the matrices always square?

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:263]: The third term involves rectangular matrices, and the order of operations which you suggested is what I had implemented. 

By stability I mean that computations that were working before i.e. CMMD(x, x) = 0 from the tests are now giving non-zero or negative answrs indicating we are having numerically instability issues where there weren'tpreviously.

File: coreax/coreset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coreset.py]: Happy to remove these from the PR.

File: coreax/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/metrics.py:263]: Ok, that all makes sense. If you want to get this PR merged ASAP, open a new ticket to revist this (potential) optimisation at a future date.

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:263]: Okay, CMMD is the original implementation, CMMD2 is the Hadamard implementation and CMMD3 is a mix of the two where we avoid the transpose in the third term by using the trace. These are the timings:

image

Clearly the Hadamard method is the fastest, and the original is the slowest, though the difference isn't huge.

Running the tests on `float32` we can see the original CMMD is the most stable (in the sense I described above):

image

Enabling `float64` results in lots of negative values of CMMD for CMMD2 and CMMD3 (resulting in `nan`s), but none for the original implementation:

image

Ignore the approximator tests, that is likely a bug with enforcing 32 bit somewhere in the code.

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:263]: I think this supports sticking with the original implementation as the speed improvements are not order of magnitudes.

[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/metrics.py:263]: I'm happy with that. If you still have the code, could you perform the benchmarks on a GPU as well. I have a feeling that the specialised matrix-multiply hardware (E.G. Nvidia's Tensor Cores) might allow the original implementation to be faster on these devices anyway.

**Review by tc85324 — APPROVED** [review]
Thanks for this, it all looks good to me!

File: coreax/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:263]: Sure, I'll set up an EC2 hopefully today and get back to you!
